robot/impl/resource_manager.go

```diff
@@ -537,115 +539,150 @@ func (manager *resourceManager) completeConfig(
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+
+					default:
+						err := errors.New("config is not for a component or service")
+						gNode.LogAndSetLastError(err, "resource", resName)
 					}
-				}
+				})
 
-				if err != nil {
-					gNode.LogAndSetLastError(
-						fmt.Errorf("resource build error: %w", err),
-						"resource", conf.ResourceName(),
-						"model", conf.Model)
-					return
+				select {
+				case <-resChan:
+				case <-ctxWithTimeout.Done():
+					if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+						lr.logger.CWarn(ctx, rutils.NewBuildTimeoutError(resName.String()))
+					}
+				case <-ctx.Done():
+					return ctx.Err()
 				}
+				return nil
+			}
 
-				// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
-				// resource generation is running async, and we don't currently have good
-				// validation around how this might affect the resource graph. So, we avoid
-				// updating the graph to be safe.
-				if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
-					manager.logger.CErrorw(
-						ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
-				} else {
-					gNode.SwapResource(newRes, conf.Model)
+			syncRes := forceSync
+			if !syncRes {
+				// TODO(RSDK-6925): support concurrent processing of resources of
+				// APIs with a maximum instance limit. Currently this limit is
+				// validated later in the resource creation flow and assumes that
+				// each resource is created synchronously to have an accurate
+				// creation count.
+				if c, ok := resource.LookupGenericAPIRegistration(resName.API); ok && c.MaxInstance != 0 {
+					syncRes = true
 				}
-
-			default:
-				err := errors.New("config is not for a component or service")
-				gNode.LogAndSetLastError(err, "resource", resName)
 			}
-		})
 
-		select {
-		case <-resChan:
-		case <-ctxWithTimeout.Done():
-			if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
-				lr.logger.CWarn(ctx, rutils.NewBuildTimeoutError(resName.String()))
+			lr.reconfigureWorkers.Add(1)
```

Comment:
This feels a bit removed from its sibling `Done`. I'd consider removing the `reconfigureWorkers.Done()` from the `processResource` and instead write:

```
			if forceSyncRes {
				if err := processResource(); err != nil {
					return
				}
			} else {
				lr.reconfigureWorkers.Add(1)
				levelErrG.Go(func() error {
					defer lr.reconfigureWorkers.Done()
					return processResource()
				})
			}
```

Reply:
nice! this is much better